src/Symfony/Component/PropertyAccess/PropertyAccessor.php

```diff
@@ -390,17 +394,22 @@ private function readProperty($zval, $property, bool $ignoreInvalidProperty = fa
 
         $result = self::$resultProto;
         $object = $zval[self::VALUE];
-        $access = $this->getReadAccessInfo(\get_class($object), $property);
+        $class = \get_class($object);
+        $access = $this->getReadAccessor($class, $property);
+
+        if (null !== $access) {
+            if (ReadAccessor::TYPE_METHOD === $access->getType()) {
+                $result[self::VALUE] = $object->{$access->getName()}();
+            }
 
-        if (self::ACCESS_TYPE_METHOD === $access[self::ACCESS_TYPE]) {
-            $result[self::VALUE] = $object->{$access[self::ACCESS_NAME]}();
-        } elseif (self::ACCESS_TYPE_PROPERTY === $access[self::ACCESS_TYPE]) {
-            $result[self::VALUE] = $object->{$access[self::ACCESS_NAME]};
+            if (ReadAccessor::TYPE_PROPERTY === $access->getType()) {
+                $result[self::VALUE] = $object->{$access->getName()};
 
-            if ($access[self::ACCESS_REF] && isset($zval[self::REF])) {
-                $result[self::REF] = &$object->{$access[self::ACCESS_NAME]};
+                if (isset($zval[self::REF]) && $access->canBeReference()) {
+                    $result[self::REF] = &$object->{$access->getName()};
+                }
             }
-        } elseif (!$access[self::ACCESS_HAS_PROPERTY] && property_exists($object, $property)) {
+        } elseif ($object instanceof \stdClass && property_exists($object, $property)) {
```

Comment:
i we could refactor this to do early return instead of all the elseif, while we are touching this

Reply:
Early return may change behavior ? If so, It's a BC break, and would prefer to not touch this

Reply:
We talk about this, and we prefer to not change this here, as it may have some undefined behavior. And would be better to do it in another PR.